File: data_structures_and_algorithm/linked_list/doubly_linked_list/insert_end.c
// insert_end.c
// This code defines a simple doubly linked list in C and provides a function to insert a new node at the end of the list.
// It also includes a function to print the linked list for demonstration purposes.

#include <stdio.h>
#include <stdlib.h>

struct Node {                       // This structure defines a node in the doubly linked list.
    int data;
    struct Node* prev;
    struct Node* next;
};

struct Node* insert_at_end(struct Node* head, int data) {                   // This function inserts a new node with the given data at the end of the list.
    struct Node* new_node = (struct Node*)malloc(sizeof(struct Node));
    new_node->data = data;
    new_node->next = NULL;

    if (head == NULL) {
        new_node->prev = NULL;
        return new_node;
    }

    struct Node* temp = head;
    while (temp->next != NULL)
        temp = temp->next;

    temp->next = new_node;
    new_node->prev = temp;

    return head;
}

void print_list(struct Node* head) {                // This function prints the linked list from head to tail.
    while (head != NULL) {
        printf("%d <-> ", head->data);
        head = head->next;
    }
    printf("NULL\n");
}

int main() {            // Example usage of the doubly linked list.
    struct Node* head = NULL;
    head = insert_at_end(head, 10);
    head = insert_at_end(head, 20);
    head = insert_at_end(head, 30);
    print_list(head);
    return 0;
}

File: data_structures_and_algorithm/linked_list/doubly_linked_list/insert_end.py
## This code is part of a simple doubly linked list implementation in Python.
## It provides functions to insert nodes at the end and print the list.

class Node:                             ## Define a Node class for the doubly linked list    
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None

def insert_at_end(head, data):          ## Function to insert a new node at the end of the doubly linked list
    new_node = Node(data)
    if head is None:
        return new_node
    temp = head
    while temp.next:
        temp = temp.next
    temp.next = new_node
    new_node.prev = temp
    return head

def print_list(head):                   ## Function to print the doubly linked list
    print("Doubly Linked List: ", end="")
    while head:
        print(head.data, end=" <-> ")
        head = head.next
    print("None")

if __name__ == "__main__":            ## Example usage of the doubly linked list
    head = None
    head = insert_at_end(head, 10)
    head = insert_at_end(head, 20)
    head = insert_at_end(head, 30)
    print_list(head)
